src/models/transfers.js

```diff
@@ -0,0 +1,354 @@
+  transfer.executed_at = updatedTransferData.executed_at
+  transfer.rejected_at = updatedTransferData.rejected_at
+
+  // Ignore undefined properties
+  const transferData = _.omit(transfer.getData(), _.isUndefined)
+
+  // Clients can add authorizations
+  // The validity of these authorizations will be checked
+  // in the validateAuthorizations function
+  _.forEach(updatedTransferData.debits, function (funds, i) {
+    if (!funds.authorized &&
+      transfer.debits[i] &&
+      transfer.debits[i].authorized) {
+      funds.authorized = true
+    }
+  })
+  _.forEach(updatedTransferData.credits, function (funds, i) {
+    if (!funds.authorized &&
+      transfer.credits[i] &&
+      transfer.credits[i].authorized) {
+      funds.authorized = true
+    }
+  })
+
+  // Clients may fulfill the execution/cancellation conditions
+  if (transfer.execution_condition_fulfillment) {
+    updatedTransferData.execution_condition_fulfillment =
+      transfer.execution_condition_fulfillment
+  }
+  if (transfer.cancellation_condition_fulfillment) {
+    updatedTransferData.cancellation_condition_fulfillment =
+      transfer.cancellation_condition_fulfillment
+  }
+
+  // The old and new objects should now be exactly equal
+  if (!_.isEqual(updatedTransferData, transferData)) {
+    // If they aren't, this means the user tried to update something they're not
+    // supposed to be able to modify.
+    // TODO InvalidTransformationError
+    throw new InvalidModificationError(
+      'Transfer may not be modified in this way',
+      diff(updatedTransferData, transferData))
+  }
+
+  originalTransfer.setData(updatedTransferData)
+  return originalTransfer
+}
+
+function validateAuthorizations (authorizedAccount, transfer, previousDebits) {
+  // Check that the authorizedAccount is actually relevant to this transfer
+  if (authorizedAccount) {
+    if (!_.includes(_.pluck(transfer.debits, 'account'), authorizedAccount) &&
+      !_.includes(_.pluck(transfer.credits, 'account'), authorizedAccount)) {
+      // TODO: should this error be more specific or would that create
+      // a security vulnerability?
+      throw new UnauthorizedError('Unknown or invalid account / password')
+    }
+  }
+
+  // Go through each of the debits in the transfer
+  // Throw an error if authorized is set to true on any debit
+  // that was not set to true on the transferFromDb and is not
+  // owned by the authorizedAccount
+  _.forEach(transfer.debits, function (debit, debitIndex) {
+    // We don't care about debits where no one is attempting
+    // to mark them as authorized
+    if (!debit.authorized) {
+      return
+    }
+
+    if (debit.account !== authorizedAccount &&
+      (!previousDebits || !previousDebits[debitIndex].authorized)) {
+      throw new UnauthorizedError('Invalid attempt to authorize debit')
+    }
+  })
+
+// TODO: add credit authorization
+}
+
+function * processStateTransitions (tr, transfer) {
```

Comment:
I guess `tr` is a db transaction, but I think it would be a lot clearer if this was spelled out (rename `tr` to `transaction`).